File: src/bridge/memoryCanvas.ts
import { z } from "zod";

import { invokeNative, voidResponse } from "./native";

// ── Schemas ──────────────────────────────────────────

const memoryCanvasSchema = z
  .object({
    id: z.string(),
    workspaceId: z.string(),
    name: z.string(),
    description: z.string().nullable(),
    nodeCount: z.number(),
    edgeCount: z.number(),
    createdAt: z.string(),
    updatedAt: z.string(),
  })
  .passthrough();

const memoryNodeSchema = z
  .object({
    id: z.string(),
    canvasId: z.string(),
    nodeType: z.string(),
    positionX: z.number(),
    positionY: z.number(),
    width: z.number().nullable(),
    height: z.number().nullable(),
    payloadJson: z.string(),
    summary: z.string().nullable(),
    assetId: z.string().nullable(),
    createdAt: z.string(),
    updatedAt: z.string(),
  })
  .passthrough();

const memoryEdgeSchema = z
  .object({
    id: z.string(),
    canvasId: z.string(),
    sourceNodeId: z.string(),
    targetNodeId: z.string(),
    edgeType: z.string(),
    label: z.string().nullable(),
    createdAt: z.string(),
  })
  .passthrough();

const memoryViewportSchema = z
  .object({
    canvasId: z.string(),
    zoom: z.number(),
    panX: z.number(),
    panY: z.number(),
    updatedAt: z.string(),
  })
  .passthrough();

const canvasListSchema = z.array(memoryCanvasSchema);
const nodeListSchema = z.array(memoryNodeSchema);
const edgeListSchema = z.array(memoryEdgeSchema);

// ── Types ────────────────────────────────────────────

export type MemoryCanvas = z.infer<typeof memoryCanvasSchema>;
export type MemoryNode = z.infer<typeof memoryNodeSchema>;
export type MemoryEdge = z.infer<typeof memoryEdgeSchema>;
export type MemoryViewport = z.infer<typeof memoryViewportSchema>;

// ── Canvas IPC ───────────────────────────────────────

export async function createCanvas(
  workspaceId: string,
  name: string,
  description?: string,
): Promise<MemoryCanvas> {
  return invokeNative("memory_canvas_v1_create", memoryCanvasSchema, {
    request: { workspace_id: workspaceId, name, description },
  });
}

export async function listCanvases(workspaceId: string): Promise<MemoryCanvas[]> {
  return invokeNative("memory_canvas_v1_list", canvasListSchema, {
    request: { workspace_id: workspaceId },
  });
}

export async function getCanvas(id: string): Promise<MemoryCanvas | null> {
  return invokeNative("memory_canvas_v1_get", memoryCanvasSchema.nullable(), {
    request: { id },
  });
}

export async function deleteCanvas(id: string): Promise<void> {
  await invokeNative("memory_canvas_v1_delete", voidResponse, {
    request: { id },
  });
}

// ── Node IPC ─────────────────────────────────────────

export async function addNode(
  canvasId: string,
  nodeType: string,
  positionX: number,
  positionY: number,
  payloadJson: string,
  summary?: string,
  assetId?: string,
): Promise<MemoryNode> {
  return invokeNative("memory_node_v1_add", memoryNodeSchema, {
    request: {
      canvas_id: canvasId,
      node_type: nodeType,
      position_x: positionX,
      position_y: positionY,
      payload_json: payloadJson,
      summary,
      asset_id: assetId,
    },
  });
}

export async function updateNode(
  id: string,
  canvasId: string,
  nodeType: string,
  positionX: number,
  positionY: number,
  payloadJson: string,
  summary?: string,
  assetId?: string,
): Promise<MemoryNode> {
  return invokeNative("memory_node_v1_update", memoryNodeSchema, {
    request: {
      id,
      canvas_id: canvasId,
      node_type: nodeType,
      position_x: positionX,
      position_y: positionY,
      payload_json: payloadJson,
      summary,
      asset_id: assetId,
    },
  });
}

export async function deleteNode(id: string): Promise<void> {
  await invokeNative("memory_node_v1_delete", voidResponse, {
    request: { id },
  });
}

export async function listNodes(canvasId: string): Promise<MemoryNode[]> {
  return invokeNative("memory_node_v1_list", nodeListSchema, {
    request: { canvas_id: canvasId },
  });
}

// ── Edge IPC ─────────────────────────────────────────

export async function addEdge(
  canvasId: string,
  sourceNodeId: string,
  targetNodeId: string,
  edgeType: string,
  label?: string,
): Promise<MemoryEdge> {
  return invokeNative("memory_edge_v1_add", memoryEdgeSchema, {
    request: {
      canvas_id: canvasId,
      source_node_id: sourceNodeId,
      target_node_id: targetNodeId,
      edge_type: edgeType,
      label,
    },
  });
}

export async function deleteEdge(id: string): Promise<void> {
  await invokeNative("memory_edge_v1_delete", voidResponse, {
    request: { id },
  });
}

export async function listEdges(canvasId: string): Promise<MemoryEdge[]> {
  return invokeNative("memory_edge_v1_list", edgeListSchema, {
    request: { canvas_id: canvasId },
  });
}

// ── Viewport IPC ─────────────────────────────────────

export async function saveViewport(
  canvasId: string,
  zoom: number,
  panX: number,
  panY: number,
): Promise<void> {
  await invokeNative("memory_viewport_v1_save", voidResponse, {
    request: { canvas_id: canvasId, zoom, pan_x: panX, pan_y: panY },
  });
}

export async function getViewport(canvasId: string): Promise<MemoryViewport> {
  return invokeNative("memory_viewport_v1_get", memoryViewportSchema, {
    request: { canvas_id: canvasId },
  });
}

File: src-tauri/src/ipc/memory_canvas.rs
use tauri::{AppHandle, Manager};

use crate::{
    application::memory_canvas_service::MemoryCanvasService,
    ports::memory_canvas_repository::{MemoryCanvas, MemoryEdge, MemoryNode, MemoryViewport},
};

use super::error::IpcError;

// ──────────────────────────────────────────────────────────────────
// Canvas commands
// ──────────────────────────────────────────────────────────────────

#[derive(serde::Deserialize)]
pub struct CreateCanvasRequest {
    pub workspace_id: String,
    pub name: String,
    pub description: Option<String>,
}

#[tauri::command]
pub async fn memory_canvas_v1_create(
    app: AppHandle,
    request: CreateCanvasRequest,
) -> Result<MemoryCanvas, IpcError> {
    let service = app.state::<MemoryCanvasService>();
    let canvas = service.create_canvas(
        &request.workspace_id,
        &request.name,
        request.description.as_deref(),
    )?;
    Ok(canvas)
}

#[derive(serde::Deserialize)]
pub struct ListCanvasesRequest {
    pub workspace_id: String,
}

#[tauri::command]
pub async fn memory_canvas_v1_list(
    app: AppHandle,
    request: ListCanvasesRequest,
) -> Result<Vec<MemoryCanvas>, IpcError> {
    let service = app.state::<MemoryCanvasService>();
    let canvases = service.list_canvases(&request.workspace_id)?;
    Ok(canvases)
}

#[derive(serde::Deserialize)]
pub struct GetCanvasRequest {
    pub id: String,
}

#[tauri::command]
pub async fn memory_canvas_v1_get(
    app: AppHandle,
    request: GetCanvasRequest,
) -> Result<Option<MemoryCanvas>, IpcError> {
    let service = app.state::<MemoryCanvasService>();
    let canvas = service.get_canvas(&request.id)?;
    Ok(canvas)
}

#[derive(serde::Deserialize)]
pub struct DeleteCanvasRequest {
    pub id: String,
}

#[tauri::command]
pub async fn memory_canvas_v1_delete(
    app: AppHandle,
    request: DeleteCanvasRequest,
) -> Result<(), IpcError> {
    let service = app.state::<MemoryCanvasService>();
    service.delete_canvas(&request.id)?;
    Ok(())
}

// ──────────────────────────────────────────────────────────────────
// Node commands
// ──────────────────────────────────────────────────────────────────

#[derive(serde::Deserialize)]
pub struct AddNodeRequest {
    pub canvas_id: String,
    pub node_type: String,
    pub position_x: f64,
    pub position_y: f64,
    pub payload_json: String,
    pub summary: Option<String>,
    pub asset_id: Option<String>,
}

#[tauri::command]
pub async fn memory_node_v1_add(
    app: AppHandle,
    request: AddNodeRequest,
) -> Result<MemoryNode, IpcError> {
    let service = app.state::<MemoryCanvasService>();
    let node = service.add_node(
        &request.canvas_id,
        &request.node_type,
        request.position_x,
        request.position_y,
        &request.payload_json,
        request.summary.as_deref(),
        request.asset_id.as_deref(),
    )?;
    Ok(node)
}

#[derive(serde::Deserialize)]
pub struct UpdateNodeRequest {
    pub id: String,
    pub canvas_id: String,
    pub node_type: String,
    pub position_x: f64,
    pub position_y: f64,
    pub payload_json: String,
    pub summary: Option<String>,
    pub asset_id: Option<String>,
}

#[tauri::command]
pub async fn memory_node_v1_update(
    app: AppHandle,
    request: UpdateNodeRequest,
) -> Result<MemoryNode, IpcError> {
    let service = app.state::<MemoryCanvasService>();
    let node = service.update_node(
        &request.id,
        &request.canvas_id,
        &request.node_type,
        request.position_x,
        request.position_y,
        &request.payload_json,
        request.summary.as_deref(),
        request.asset_id.as_deref(),
    )?;
    Ok(node)
}

#[derive(serde::Deserialize)]
pub struct DeleteNodeRequest {
    pub id: String,
}

#[tauri::command]
pub async fn memory_node_v1_delete(
    app: AppHandle,
    request: DeleteNodeRequest,
) -> Result<(), IpcError> {
    let service = app.state::<MemoryCanvasService>();
    service.delete_node(&request.id)?;
    Ok(())
}

#[derive(serde::Deserialize)]
pub struct ListNodesRequest {
    pub canvas_id: String,
}

#[tauri::command]
pub async fn memory_node_v1_list(
    app: AppHandle,
    request: ListNodesRequest,
) -> Result<Vec<MemoryNode>, IpcError> {
    let service = app.state::<MemoryCanvasService>();
    let nodes = service.list_nodes(&request.canvas_id)?;
    Ok(nodes)
}

// ──────────────────────────────────────────────────────────────────
// Edge commands
// ──────────────────────────────────────────────────────────────────

#[derive(serde::Deserialize)]
pub struct AddEdgeRequest {
    pub canvas_id: String,
    pub source_node_id: String,
    pub target_node_id: String,
    pub edge_type: String,
    pub label: Option<String>,
}

#[tauri::command]
pub async fn memory_edge_v1_add(
    app: AppHandle,
    request: AddEdgeRequest,
) -> Result<MemoryEdge, IpcError> {
    let service = app.state::<MemoryCanvasService>();
    let edge = service.add_edge(
        &request.canvas_id,
        &request.source_node_id,
        &request.target_node_id,
        &request.edge_type,
        request.label.as_deref(),
    )?;
    Ok(edge)
}

#[derive(serde::Deserialize)]
pub struct DeleteEdgeRequest {
    pub id: String,
}

#[tauri::command]
pub async fn memory_edge_v1_delete(
    app: AppHandle,
    request: DeleteEdgeRequest,
) -> Result<(), IpcError> {
    let service = app.state::<MemoryCanvasService>();
    service.delete_edge(&request.id)?;
    Ok(())
}

#[derive(serde::Deserialize)]
pub struct ListEdgesRequest {
    pub canvas_id: String,
}

#[tauri::command]
pub async fn memory_edge_v1_list(
    app: AppHandle,
    request: ListEdgesRequest,
) -> Result<Vec<MemoryEdge>, IpcError> {
    let service = app.state::<MemoryCanvasService>();
    let edges = service.list_edges(&request.canvas_id)?;
    Ok(edges)
}

// ──────────────────────────────────────────────────────────────────
// Viewport commands
// ──────────────────────────────────────────────────────────────────

#[derive(serde::Deserialize)]
pub struct SaveViewportRequest {
    pub canvas_id: String,
    pub zoom: f64,
    pub pan_x: f64,
    pub pan_y: f64,
}

#[tauri::command]
pub async fn memory_viewport_v1_save(
    app: AppHandle,
    request: SaveViewportRequest,
) -> Result<(), IpcError> {
    let service = app.state::<MemoryCanvasService>();
    service.save_viewport(
        &request.canvas_id,
        request.zoom,
        request.pan_x,
        request.pan_y,
    )?;
    Ok(())
}

#[derive(serde::Deserialize)]
pub struct GetViewportRequest {
    pub canvas_id: String,
}

#[tauri::command]
pub async fn memory_viewport_v1_get(
    app: AppHandle,
    request: GetViewportRequest,
) -> Result<MemoryViewport, IpcError> {
    let service = app.state::<MemoryCanvasService>();
    let viewport = service.get_viewport(&request.canvas_id)?;
    Ok(viewport)
}
